### PR TITLE
WIP: Replace foreman with Jenkins based backend

### PR DESCRIPTION
**Pull request by olivergondza**
@pjanouse 

TODO:
- [x] Rename modules
- [x] Rename repository
- [x] Extract shared code for both plugins
- [x] Create integration test module based on JTH
- [x] Fix Jenkinsfile
- [x] Fix container to use Jenkins/backend (do we still need the container?)
- [x] Fix Maven version - set to 2.0
- [x] Rename package for plugin to com.redhat.jenkins.nodesharingfrontend
- [x] Remove all occurrences of Foreman in the code, readme etc.
- [x] Fix project url in pom.xml
- [x] Authentication
- [x] ACL check on REST methods
- [x] Test with crumb issuer
- [ ] Cloud statistics on executor side
- [ ] Clean Messages files of unused strings

Test scenarios:
- [ ] Orchestrator restart
- [ ] Orchestrator URL is changed and distributed via config repo
- [ ] Config repository URL change
- [ ] Reject requests from executors not in config repo
- [ ] Deactivate executor not in config repo

After Merge:
- [ ] Rename/update wiki page to https://wiki.jenkins-ci.org/display/JENKINS/Node+Sharing+Plugin
- [ ] Remove old plugin from distribution: https://github.com/jenkins-infra/backend-update-center2/blob/master/src/main/resources/artifact-ignores.properties
- [ ] Grant release permission to renamed components: https://github.com/jenkins-infra/repository-permissions-updater/pull/342

### CODE REVIEW

File: plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java [hunk 1]
```diff
@@ -96,6 +99,13 @@ public Api(String OrchestratorUrl) {
         base = client.target(OrchestratorUrl);
     }
 
+    private SharedNodeCloud getCloud() {
+        if (cloud == null) {
+            throw new ActionFailed.CommunicationError("Cloud isn't set!");
```
Comment (olivergondza) [inline plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java]: Is there a reason to ever construct an `Api` object without it? Provided it is not, let's enforce that in constructor.

File: plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java [hunk 2]
```diff
@@ -219,14 +229,10 @@ public Response doPostRequest(@Nonnull final WebTarget target, @Nonnull final St
...
     @CheckForNull
     public Object nodeStatus(@Nonnull @QueryParameter("name") final String name) {
-        // TODO Response to node status
-        Communication.NodeState status = Communication.NodeState.NOT_FOUND;
-        Node node = Jenkins.getActiveInstance().getNode(name);
-        if (node != null) {
-            // TODO Extract the current states
-            status = Communication.NodeState.FOUND;
+        if(name == null) {
+            throw new ActionFailed.CommunicationError("Node name cannot be 'null'!");
```
Comment (olivergondza) [inline plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java]: This is an API abuse - `IllegalArgumentException`.

File: plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java (hunk 1 above)
Comment (pjanouse) [reply plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java]: Ok, but we need a getApi() for testing purposes anyway...

File: plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java (hunk 2 above)
Comment (pjanouse) [reply plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java]: :+1:

File: plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java (hunk 1 above)
Comment (pjanouse) [reply plugin/src/main/java/com/redhat/jenkins/nodesharingfrontend/Api.java]: BTW in that case we need to deal with Interrupted ex. in the constructor... how we should handle it there? Set api var to null?

File: jth-tests/src/test/java/com/redhat/jenkins/nodesharing/ConfigRepoRule.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        repos.add(repo);
+
+        StreamTaskListener listener = new StreamTaskListener(System.err, Charset.defaultCharset());
+        GitClient git = Git.with(listener, new EnvVars()).in(repo).using("git").getClient();
+        git.init();
+        git.add("*");
+        git.commit("Init");
+
+        FilePath config = git.getWorkTree().child("config");
+        // TODO replase Jenkins URL
```
Comment (olivergondza) [inline jth-tests/src/test/java/com/redhat/jenkins/nodesharing/ConfigRepoRule.java]: Are these TODOs still relevant?

File: jth-tests/src/test/java/com/redhat/jenkins/nodesharing/ConfigRepoRule.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        String newConfig = config.readToString().replace("orchestrator.url=",
+                "orchestrator.url=" + j.getRootUrl());
+        config.write(newConfig, Charset.defaultCharset().name());
+        git.add("config");
+
+        FilePath jenkinses = git.getWorkTree().child("jenkinses");
+        // TODO replase Jenkins URL
```
Comment (olivergondza) [inline jth-tests/src/test/java/com/redhat/jenkins/nodesharing/ConfigRepoRule.java]: Ditto?

File: jth-tests/src/test/java/com/redhat/jenkins/nodesharing/ConfigRepoRule.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        String newJenkinses = jenkinses.readToString().replace("jenkins1=",
+                "jenkins1=" + j.getRootUrl());
+        jenkinses.write(newJenkinses, Charset.defaultCharset().name());
+        git.add("jenkinses");
+
+        // TODO create and store real Slave
```
Comment (olivergondza) [inline jth-tests/src/test/java/com/redhat/jenkins/nodesharing/ConfigRepoRule.java]: Ditto?

File: jth-tests/src/test/resources/com/redhat/jenkins/nodesharing/dummy_config_repo/nodes/solaris2.acme.com.xml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 <!--TODO: Fix the name duality-->
 <com.redhat.jenkins.nodesharingfrontend.SharedNode>
-  <name>solaris2.executor.com</name>
+  <name>solaris2.acme.com</name>
```
Comment (olivergondza) [inline jth-tests/src/test/resources/com/redhat/jenkins/nodesharing/dummy_config_repo/nodes/solaris2.acme.com.xml]: @pjanouse, FYI

**Pull request merged**